flang/lib/Lower/ConvertVariable.cpp

```diff
@@ -483,6 +592,78 @@ static std::string mangleGlobalAggregateStore(
   assert(st.isGlobal() && "cannot name local aggregate");
   return Fortran::lower::mangle::mangleName(*st.vars[0]);
 }
+
+// In equivalences, 19.5.3.4 point 10 gives rules regarding explicit and default
+// initialization interaction. The interpretation of this point is not
+// ubiquitous among compilers, but the majority (nag, ifort, nvfortran) accept
+// full or partial overlap of explicit init over default init. In case of
+// partial overlap, these compiler still apply the default init for the
+// components for which the storage is not explicitly initialized.
+//
+// In the example below, the equivalence storage must be initialized to [3 , 2].
+// 3 comes from the explicit init of k, and 2 from part of the deafult init of
+// a.
+//
+//  type t
+//    i = 1
+//    j = 2
+//  end type
+//  type(t), save :: a
+//  integer, save :: k = 3
+//  equivalence (a, k)
+//
+
+/// Helper to analyze overlaps between default and explicit initialization in an
+/// equivalences. \p offset is the current equivalence offset, memIter is the
+/// current symbol iterator over the equivalence members. It must point to a
+/// member that has default initialization. \p endIter must be the end of the
+/// equivalence members iterators. This will tell if the equivalence member \p
+/// memIter is fully, partially, or not overlapped by members with explicit
+/// initialization.
```

Comment:
Do we have to consider overlapping (and conflicting) default initialization cases? Or possibly just have some asserts to make sure they are caught by semantics?

e.g.,
```Fortran
type u
  integer :: pad
  type(t) :: foo
end type u

type(u) :: b
equivalence (a, b)
```
so b%foo%i is the same location as a%j, and they have to arm wrestle for it.